azure/services/async/async.go

```diff
@@ -183,12 +190,40 @@ func (s *Service) DeleteResource(ctx context.Context, spec azure.ResourceSpecGet
 	return nil
 }
 
-// retryAfter returns the max between the `RETRY-AFTER` header and the default requeue time.
+// getRequeueAfterFromFuture returns the max between the `RETRY-AFTER` header and the default requeue time.
 // This ensures we respect the retry-after header if it is set and avoid retrying too often during an API throttling event.
-func retryAfter(sdkFuture azureautorest.FutureAPI) time.Duration {
+func getRequeueAfterFromFuture(sdkFuture azureautorest.FutureAPI) time.Duration {
 	retryAfter, _ := sdkFuture.GetPollingDelay()
 	if retryAfter < reconciler.DefaultReconcilerRequeue {
 		retryAfter = reconciler.DefaultReconcilerRequeue
 	}
 	return retryAfter
 }
+
+// getRetryAfterFromError returns the time.Duration from the http.Response in the autorest.DetailedError.
+// If there is no Response object, or if there is no meaningful Retry-After header data, we return a default.
+func getRetryAfterFromError(err error) (ret time.Duration) {
+	// In case we aren't able to introspect Retry-After from the error type, we'll return this default
+	ret = reconciler.DefaultReconcilerRequeue
+	var detailedError autorest.DetailedError
+	// if we have a strongly typed autorest.DetailedError then we can introspect the HTTP response data
+	if errors.As(err, &detailedError) {
+		if detailedError.Response != nil {
+			// If we have Retry-After HTTP header data for any reason, prefer it
+			if retryAfter := detailedError.Response.Header.Get("Retry-After"); retryAfter != "" {
+				// This handles the case where Retry-After data is in the form of units of seconds
```

Comment:
are we expecting both scenarios to be possible (units of seconds and absolute time)? Same question with `autorest.DetailedError` vs not ? Are there any SDK contracts for those error responses / headers that we can follow? Seems strange that we have to handle multiple possibilities, it's like we're guessing what the reponse might look like instead of expecting a specific format/unit/type.

Reply:
> are we expecting both scenarios to be possible (units of seconds and absolute time)

As far as I can tell from research, the spec is overloaded to expect both value type flavors, so I think it's the best practice to deal with both types wherever we parse `Retry-After` HTTP header data.

> Same question with autorest.DetailedError vs not

The `autorest.DetailedError` error implementation is one that we definitely know about from the usage above in the async flow, based on the specific SDK API we're currently re-using. In that sense you could say that this helper function is sort of tightly coupled to the particular implementation of capz at this point in time (initially this foo was inline in the `CreateResource` func but I split it out for code maintenance reasons). So tl;dr we only have access to `Retry-After` data in this particular case via the `err` response from the underlying service `Get` implementations, and we only know how to get at it (as of right now) if the err is of "type" `autorest.DetailedError`. Hope that makes sense!